File: c++/tests/test_parseProcess.cpp
#include <vector>

#include "catch.hpp"

// Include the cpp file we're testing
#include "shell_utils.cpp"
#include "process.h"

TEST_CASE( "Parse: One Command - Zero Args", "[classic]") {
    GIVEN("Parse: One Command - Zero Args"){
        std::string commandString = "du";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 1 );
        CHECK( processes[0].numArguments() == 0 );
    }
}

TEST_CASE( "Parse: One Command - Zero Args With Empty String", "[classic]") {
    GIVEN("Parse: One Command - Zero Args With Empty String"){
        std::string commandString = "du \"\"";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 1 );
        CHECK( processes[0].numArguments() == 0 );
    }
}

TEST_CASE( "Parse: One Command - One Args", "[classic]") {
    GIVEN("Parse: One Command - One Args"){
        std::string commandString = "du -l";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 1 );
        CHECK( processes[0].numArguments() == 1 );
    }
}

TEST_CASE( "Parse Counting: Two Command With Spaces", "[classic]") {
    GIVEN("Parse: Two Command With Spaces"){
        std::string commandString = "du -l | ps aux";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 2 );
    }
}

TEST_CASE( "Parse Counting: Two Commands With No Spaces", "[classic]") {
    GIVEN("Parse: Two Commands With No Spaces"){
        std::string commandString = "du -l|ps aux";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 2 );
    }
}

TEST_CASE( "Parse Counting: Two Commands With Mixed Spaces", "[classic]") {
    GIVEN("Parse: Two Commands With Mixed Spaces"){
        std::string commandString = "du -l| ps aux";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 2 );
    }
}

TEST_CASE( "Parse Counting: Three Commands With Mixed Spaces", "[classic]") {
    GIVEN("Parse: Three Commands With Mixed Spaces"){
        std::string commandString = "du -l| ps aux | grep argument";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}

TEST_CASE( "Parse Counting: Three Commands With Spaces Zero Args", "[classic]") {
    GIVEN("Parse: Three Commands With Spaces Zero Args"){
        std::string commandString = "du | ps | grep";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}

TEST_CASE( "Parse Counting: Three Commands No Spaces One Arg", "[classic]") {
    GIVEN("Parse: Three Commands With No Spaces One Arg"){
        std::string commandString = "du -l| ps aux| grep argument";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}

TEST_CASE( "Parse Counting: Three Commands Mixed Spaces One Arg", "[classic]") {
    GIVEN("Parse: Three Commands With Mixed Spaces One Arg"){
        std::string commandString = "du -l | ps aux |grep argument";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}

TEST_CASE( "Parse Counting: Three Commands Trailing Spaces One Arg", "[classic]") {
    GIVEN("Parse: Three Commands With Trailing Spaces One Arg"){
        std::string commandString = "du -l |ps aux |grep argument   ";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}
    
TEST_CASE( "Parse Counting: Three Commands Leading Spaces One Arg", "[classic]") {
    GIVEN("Parse: Three Commands With Leading Spaces One Arg"){
        std::string commandString = "   du -l |ps aux |grep argument";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}

TEST_CASE( "Parse Counting: Three Commands Mixed Spaces Less Than Greater", "[classic]") {
    GIVEN("Parse Counting: Three Commands Mixed Spaces Less Than Greater"){
        std::string commandString = "du -l <ps aux >grep argument";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}
    
TEST_CASE( "Parse Counting: Three Commands Mixed Spaces Greater Than Less", "[classic]") {
    GIVEN("Parse Counting: Three Commands Mixed Spaces Greater Than Less"){
        std::string commandString = "du -l >ps aux <grep argument";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes.size() == 3 );
    }
}

// Verify Correct Commands in Parsed Processes
TEST_CASE( "Parse Commands: One Command One Arg", "[classic]") {
    GIVEN("Parse Commands: One Command One Arg"){
        std::string commandString = "du -l";
        std::vector<Process> processes = parseString(commandString);
        CHECK( processes[0].getCommand() == "du" );
    }
}

TEST_CASE( "Parse Commands: Three Commands One Arg", "[classic]") {
    GIVEN("Parse Commands: Three Commands One Arg"){
        std::string commandString = "du -l | ps aux | grep argument";
        std::vector<Process> processes = parseString(commandString);

        std::string correctCommands[] = {"du", "ps", "grep"};
        for( std::vector<Process>::size_type i = 0; i < processes.size(); i++ ){
            CHECK( processes[i].getCommand() == correctCommands[i] );
        }
    }
}

// Verify correct arguments in parsed processes
TEST_CASE( "Parse Arguments: One Command Two Arg", "[classic]") {
    GIVEN("Parse Arguments: One Command Two Arg"){
        std::string commandString = "du -l apple";
        std::vector<Process> processes = parseString(commandString);

        std::string correctArguments[][5] = {
            {"-l", "apple", "END_ARGS"}
        };
        for( std::vector<Process>::size_type i = 0; i < processes.size(); i++ ){
            std::string *arg = correctArguments[i];
            for( int j = 0; *arg != "END_ARGS"; j++ ) {
                CHECK( processes[i].getArgument(j) == *arg );
                arg++;
            }
        }
    }
}

TEST_CASE( "Parse Arguments: Three Command Variable Arg", "[classic]") {
    GIVEN("Parse Arguments: Three Command Variable Arg"){
        std::string commandString = "du -l | ps aux apple | grep argument";
        std::vector<Process> processes = parseString(commandString);

        std::string correctArguments[][5] = {
            {"-l", "END_ARGS"},
            {"aux", "apple", "END_ARGS"},
            {"argument", "END_ARGS"}
        };
        for( std::vector<Process>::size_type i = 0; i < processes.size(); i++ ){
            std::string *arg = correctArguments[i];
            for( int j = 0; *arg != "END_ARGS"; j++ ) {
                CHECK( processes[i].getArgument(j) == *arg );
                arg++;
            }
        }
    }
}

TEST_CASE( "Parse Arguments: Three Command Zero Arg", "[classic]") {
    GIVEN("Parse Arguments: Three Command Zero Arg"){
        std::string commandString = "du | ps | grep";
        std::vector<Process> processes = parseString(commandString);

        for( std::vector<Process>::size_type i = 0; i < processes.size(); i++ ){
            CHECK( processes[i].getArguments().size() == 0 );
        }
    }
}

// Cause Exceptions To Occur Via Mismatched Quotes
TEST_CASE( "MisMatch Quotes: One Command", "[classic]") {
    GIVEN("MisMatch Quotes: One Command"){
        std::string commandString = "du \"mismatch";
        REQUIRE_THROWS_AS(parseString(commandString), std::out_of_range);
    }
}

TEST_CASE( "MisMatch Quotes: Two Commands", "[classic]") {
    GIVEN("MisMatch Quotes: Two Commands"){
        std::string commandString = "du \"mismatch | welcome";
        REQUIRE_THROWS_AS(parseString(commandString), std::out_of_range);
    }
}

TEST_CASE( "MisMatch Quotes: Two Commads Three quotes", "[classic]") {
    GIVEN("MisMatch Quotes: Two Commads Three quotes"){
        std::string commandString = "du \"mismatch | welcome\" | ps welcome \"bob";
        REQUIRE_THROWS_AS(parseString(commandString), std::out_of_range);
    }
}

File: python/tests/test_parseProcess.py
import unittest
from shell import parseProcesses

class ParseProcessTester(unittest.TestCase):

    # Verify Correct Number of Processes Parsed
    def test_countProcesses_1_noArgs(self):
        processes = parseProcesses("du")
        self.assertEqual(processes[0].command, "du")
        self.assertEqual(len(processes), 1)
        self.assertEqual(len(processes[0].arguments), 0)

    def test_countProcesses_1_noArgs_emptyString(self):
        processes = parseProcesses("du \"\"")
        self.assertEqual(len(processes), 1)
        self.assertEqual(len(processes[0].arguments), 0)

    def test_countProcesses_1(self):
        processes = parseProcesses("du -l")
        self.assertEqual(len(processes), 1)
        self.assertEqual(len(processes[0].arguments), 1)

    def test_countProcesses_2_withSpaces(self):
        processes = parseProcesses('du -l | ps aux')
        self.assertEqual(len(processes), 2)

    def test_countProcesses_2_withNoSpaces(self):
        processes = parseProcesses('du -l|ps aux')
        self.assertEqual(len(processes), 2)

    def test_countProcesses_2_withMixedSpaces(self):
        processes = parseProcesses('du -l| ps aux')
        self.assertEqual(len(processes), 2)
    
    def test_countProcesses_3_withSpaces(self):
        processes = parseProcesses('du -l | ps aux | grep argument')
        self.assertEqual(len(processes), 3)

    def test_countProcesses_3_withSpacesNoArgs(self):
        processes = parseProcesses('du | ps | grep')
        self.assertEqual(len(processes), 3)

    def test_countProcesses_3_withNoSpaces(self):
        processes = parseProcesses('du -l|ps aux|grep argument')
        self.assertEqual(len(processes), 3)
    
    def test_countProcesses_3_withMixedSpaces(self):
        processes = parseProcesses('du -l |ps aux |grep argument')
        self.assertEqual(len(processes), 3)

    def test_countProcesses_3_tailingSpaces(self):
        processes_1 = parseProcesses('du -l |ps aux |grep argument   ')
        self.assertEqual(len(processes_1), 3)
    
    def test_countProcesses_3_leadingSpaces(self):
        processes_1 = parseProcesses('   du -l |ps aux |grep argument')
        self.assertEqual(len(processes_1), 3)

    def test_countProcesses_3_lessThanGreaterThan(self):
        processes_1 = parseProcesses('du -l <ps aux >grep argument')
        self.assertEqual(len(processes_1), 3)

    def test_countProcesses_3_middleManySpaces(self):
        processes_1 = parseProcesses('   du -l |ps        aux |grep argument')
        self.assertEqual(len(processes_1), 3)

    def test_countProcesses_3_greaterThanLessThan(self):
        processes_1 = parseProcesses('du -l > ps aux <grep argument')
        self.assertEqual(len(processes_1), 3)

    # Verify correct commands in parsed processes
    def test_commands_1(self):
        processes = parseProcesses('du -l')
        self.assertEqual(processes[0].command, 'du')
    
    def test_commands_3(self):
        processes = parseProcesses('du -l | ps aux | grep argument')
        
        correctCommands = ['du', 'ps', 'grep']
        for process in enumerate(processes):
            self.assertEqual(process[1].command, correctCommands[process[0]], 'Command {} incorrect'.format(process[0]))

    # Verify correct arguments in parsed processes
    def test_arguments_1(self):
        processes = parseProcesses('du -l apple')
        
        correctArguments = [
            ['-l', 'apple']
        ]
        for process in enumerate(processes):
            for argument in enumerate(process[1].arguments):
                self.assertEqual(process[1].arguments[argument[0]], correctArguments[process[0]][argument[0]], 'Command {} Argument {} incorrect'.format(process[0], argument[0]))

    def test_arguments_3Process_varyNumberOfArgs(self):
        processes = parseProcesses('du -l | ps aux apple | grep argument')
        
        correctArguments = [
            ['-l'],
            ['aux', 'apple'],
            ['argument']
        ]
        for process in enumerate(processes):
            for argument in enumerate(process[1].arguments):
                self.assertEqual(process[1].arguments[argument[0]], correctArguments[process[0]][argument[0]], 'Command {} Argument {} incorrect'.format(process[0], argument[0]))

    def test_arguments_3Process_zeroArgs(self):
        processes = parseProcesses('du | ps | grep')
        
        for process in enumerate(processes):
            self.assertEqual(len(process[1].arguments), 0, 'Command {} should have zero arguments'.format(process[0]))

    # Cause Mismatched Exceptions to occur
    def test_mismatch_quotes(self):
        self.assertRaises(Exception, parseProcesses, 'du "mismatch')
    
    def test_mismatch_quotes_2(self):
        self.assertRaises(Exception, parseProcesses, 'du "mismatch | welcome')

    def test_mismatch_quotes_3(self):
        self.assertRaises(Exception, parseProcesses, 'du "mismatch | welcome" | ps welcome "bob')